### PR TITLE
Feat: add dex login in apiserver

### PR DESCRIPTION
**Pull request by FogDong**
Signed-off-by: FogDong <dongtianxin.tx@alibaba-inc.com>

### Description of your changes

Fixes #

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### Special notes for your reviewer

/cc @barnettZQG

### CODE REVIEW

File: cmd/apiserver/main.go [hunk 1]
```diff
@@ -46,6 +47,8 @@ func main() {
 	flag.StringVar(&s.restCfg.LeaderConfig.LockName, "lock-name", "apiserver-lock", "the lease lock resource name")
 	flag.DurationVar(&s.restCfg.LeaderConfig.Duration, "duration", time.Second*5, "the lease lock resource name")
 	flag.DurationVar(&s.restCfg.AddonCacheTime, "addon-cache-duration", time.Minute*10, "how long between two addon cache operation")
+	flag.StringVar(&usecase.DexIssuerURL, "dex-issuer-url", "https://kubevela-dex-issuer.com", "the issuer url of dex")
```
Comment (barnettZQG) [inline cmd/apiserver/main.go:50]: This URL be called from vela-apiserver or browser? 

For browser,  we will proxy by VelaUX. such as "http://velaux.public/dex"

For vela-apiserver, it's better that directly through the service name.

File: pkg/apiserver/model/user.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// TableName return custom table name
+func (u *User) TableName() string {
+	return tableNamePrefix + "user"
+}
+
+// PrimaryKey return custom primary key
+func (u *User) PrimaryKey() string {
+	return strings.ReplaceAll(u.Email, "@", "_")
```
Comment (barnettZQG) [inline pkg/apiserver/model/user.go:42]:
```suggestion
	return u.Name
```

File: pkg/apiserver/rest/usecase/authentication.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	return &dexHandlerImpl{
+		idToken: idToken,
+		ds:      a.ds,
+	}, nil
+}
+
+func (a *authenticationUsecaseImpl) Login(ctx context.Context, req *restful.Request) (*apisv1.LoginResponse, error) {
+	var handler authHandler
+	var err error
+	loginType := req.QueryParameter("type")
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/authentication.go:89]: The login type should be configured by Integration, can not be defined by the requesting client.

File: pkg/apiserver/rest/usecase/authentication.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+type authHandler interface {
+	login(ctx context.Context) (*apisv1.LoginResponse, error)
+}
+
+type dexHandlerImpl struct {
+	idToken *oidc.IDToken
+	ds      datastore.DataStore
+}
+
+func (a *authenticationUsecaseImpl) newDexHandler(ctx context.Context, req *restful.Request) (*dexHandlerImpl, error) {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/authentication.go:69]: This function may be is used to validate request? It is not a login handler.

File: pkg/apiserver/rest/usecase/authentication.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+type dexHandlerImpl struct {
+	token   *oauth2.Token
+	idToken *oidc.IDToken
+	ds      datastore.DataStore
+}
+
+func (a *authenticationUsecaseImpl) newDexHandler(ctx context.Context, req *restful.Request) (*dexHandlerImpl, error) {
+	var err error
+	if dexConfig == nil {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/authentication.go:84]: The dex config maybe will be changed, so, the cache will cause an error.

**Review by wonderflow — APPROVED** [review]

**Review by yangsoon — APPROVED** [review]

**Review by leejanee — APPROVED** [review]

**Pull request merged**